src/snowflake/snowpark/modin/plugin/_internal/join_utils.py

```diff
@@ -1154,6 +1154,29 @@ def join_on_index_columns(
     return JoinOrAlignInternalFrameResult(joined_frame, result_column_mapper)
 
 
+def convert_index_type_to_variant(
+    frame: InternalFrame,
+) -> InternalFrame:
+    """
+    Converts types of given index identifier if it is not Variant or Timestamp type.
+    Args:
+        frame: InternalFrame whose type needs to be converted.
+    Returns:
+        Frame with updated columns.
+    """
+    frame_ids = frame.data_column_snowflake_quoted_identifiers
+    frame_id_to_type_map = frame.quoted_identifier_to_snowflake_type(frame_ids)
+    frame_to_variant = {}
+    for frame_id in frame_ids:
+        frame_type = frame_id_to_type_map[frame_id]
+        if frame_type not in [VariantType, TimestampType]:
```

Comment:
Why not do this for timestamp as well?

Reply:
We can actually do this for timestamp too, I added a test to confirm this. The case that doesn't work (which has negative tests already) is reindexing a datetime type frame with a non-matching index type.